### PR TITLE
Add Helm Deployment Functionality

### PR DESCRIPTION
**Pull request by kenny-chalupa**
This PR adds the ability to deploy the CDF Fabric Replicator Docker through Helm to a  Kubernetes Cluster.

## Addition of Helm Charts
- These Charts are utilized to configure the deployment of the Docker image to Kubernetes
- We introduced a config map to expose these configs to the pod at runtime as environment variables
- We introduced the client secret as a Kubernetes Secret for client authentication for the replicator at run-time
    - These Configs are used to point to required functionality such as Sink Tables and CDF Sources

See the Readme for instructions on how to deploy to Kubernetes from these Helm Charts

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -129,21 +129,140 @@ Finally, run the replicator:
...
 
-The included ".vscode/launch.json" file will add "poetry install" and "poetry run" to your "Run and Debug" tab. They can be used in place of the manual command line statements above, and will attach the VSCode debugger.
+To create an AKS cluster with Managed Identity and an attached ACR, you can use the Azure CLI:
 
-For more information see [Debugging in Visual Studio Code](https://code.visualstudio.com/Docs/editor/debugging).
+```bash
+az aks create -g MyResourceGroup -n MyManagedCluster --generate-ssh-keys --attach-acr MyACR --enable-managed-identity
+```
+
+## Build Docker to Push to ARC
```
Comment (muralidharm) [inline README.md:162]: Typo: ARC instead of ACR

File: README.md [hunk 2]
```diff
@@ -129,21 +129,140 @@ Finally, run the replicator:
...
 docker run -i -t <image-name> --platform linux/arm64 
 ```
-=======
-## Visual Studio Code
 
-Alternatively, if you are using Visual Studio Code, just open the folder of the root directory of your project.
+## Creating an AKS(Azure Kubernetes Service) Cluster with Managed Identity and Attached ACR(Azure Container Registry)
 
-You must still install Poetry as mentioned above in addition to the [Python Extension](https://marketplace.visualstudio.com/items?itemName=ms-python.python) for VSCode.
+Azure Kubernetes Service (AKS) can use Azure Managed Identities to interact with other Azure services. This eliminates the need to manage service principals and rotate credentials. The Fabric Replicator requires managed identity on AKS to be enabled to run.
```
Comment (muralidharm) [inline README.md:154]: Perhaps add this in a Requirements section with anything else needed to run the replicator (i.e. CDF, Fabric)

File: .env.example [hunk 3]
```diff
@@ -8,8 +8,6 @@ COGNITE_BASE_URL=
 COGNITE_STATE_DB=
 COGNITE_STATE_TABLE=
 COGNITE_EXTRACTION_PIPELINE=
-DPS_TABLE_NAME=
```
Comment (muralidharm) [inline .env.example:11]: Why were these removed?  There still appear to be hard coded table values in the example_config, are they supposed to be hard coded?

File: cdf-fabric-replicator-chart/templates/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                  name: cdf-conn-config
+                  key: COGNITE_TOKEN_URL
+            - name: COGNITE_CLIENT_ID
+              valueFrom:
+                configMapKeyRef:
+                  name: cdf-conn-config
+                  key: COGNITE_CLIENT_ID
+            - name: COGNITE_CLIENT_SECRET
+              valueFrom:
+                secretKeyRef:
```
Comment (muralidharm) [inline cdf-fabric-replicator-chart/templates/deployment.yaml:48]: When I last used `secretKeyRef`, the way I used it was that I created a secret using `kubectl create secret` on the kubernetes cluster then was able to reference a key in the secret in the helm chart.  Is that how this works here?

File: cdf-fabric-replicator-chart/values.yaml [hunk 5]
```diff
@@ -0,0 +1,68 @@
+# Default values for cdf-fabric-replicator-chart.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+image:
+  repository:
+  pullPolicy: Always
+  # Overrides the image tag whose default is the chart appVersion.
+  tag: "latest"
```
Comment (xtophs) [inline cdf-fabric-replicator-chart/values.yaml:9]: should we go with help suggested practices for versions: https://helm.sh/docs/chart_best_practices/dependencies/

**Review by xtophs — CHANGES_REQUESTED** [review]

File: cdf-fabric-replicator-chart/templates/deployment.yaml (hunk 4 above)
Comment (kenny-chalupa) [reply cdf-fabric-replicator-chart/templates/deployment.yaml:48]: That is one way you can initially inject the secret we create the secret in the secret yaml file and that is seeded by the values.yaml

File: README.md (hunk 2 above)
Comment (kenny-chalupa) [reply README.md:154]: We could but that would be something else we need to have a config for and manage that seemed out of scope for his story and PR

File: .env.example (hunk 3 above)
Comment (kenny-chalupa) [reply .env.example:11]: Yeah we need to introduce these back in that was probably done in the middle of moving stuff around in the config and reverting back to some of the old stuff. I will introduce these back in

Comment (kenny-chalupa) [reply .env.example]: Done

File: cdf-fabric-replicator-chart/values.yaml (hunk 5 above)
Comment (kenny-chalupa) [reply cdf-fabric-replicator-chart/values.yaml]: good point removing latest

File: README.md (hunk 2 above)
Comment (kenny-chalupa) [reply README.md:154]: I don't mind as long as its as simple as putting it all in a script with the variables such as the ACR Name the AKS Cluster Name and the App Name

File: README.md [hunk 6]
```diff
@@ -129,21 +129,140 @@ Finally, run the replicator:
...
 
-The included ".vscode/launch.json" file will add "poetry install" and "poetry run" to your "Run and Debug" tab. They can be used in place of the manual command line statements above, and will attach the VSCode debugger.
+To create an AKS cluster with Managed Identity and an attached ACR, you can use the Azure CLI:
 
-For more information see [Debugging in Visual Studio Code](https://code.visualstudio.com/Docs/editor/debugging).
+```bash
+az aks create -g MyResourceGroup -n MyManagedCluster --generate-ssh-keys --attach-acr MyACR --enable-managed-identity
```
Comment (xtophs) [inline README.md:159]: Should we also enable user-assigned managed identities in addition to the system-assigned identities?

File: cdf-fabric-replicator-chart/values.yaml [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  tag: "1.0.0"
+
+replicaCount: 1
+imagePullSecrets: []
+nameOverride: ""
+fullnameOverride: ""
+
+podAnnotations: {}
+
+podSecurityContext: {}
```
Comment (xtophs) [inline cdf-fabric-replicator-chart/values.yaml:18]: what's a secure default value here?

File: cdf-fabric-replicator-chart/values.yaml [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+securityContext: {}
```
Comment (xtophs) [inline cdf-fabric-replicator-chart/values.yaml:20]: what's a secure default value here?

File: cdf-fabric-replicator-chart/values.yaml (hunk 7 above)
Comment (kenny-chalupa) [reply cdf-fabric-replicator-chart/values.yaml:18]: Addressed  this along with Security Context here: 

https://github.com/cognitedata/cdf-fabric-replicator/pull/84#discussion_r1586504884

File: README.md (hunk 6 above)
Comment (kenny-chalupa) [reply README.md:159]: We were only using system assigned right now and for the time being we could keep that option closed off until we use them

File: README.md (hunk 2 above)
Comment (kenny-chalupa) [reply README.md:156]: Added the Pre-Requisite Section

File: .env.example [hunk 9]
```diff
@@ -8,8 +8,6 @@ COGNITE_BASE_URL=
 COGNITE_STATE_DB=
```
[the remark this answers is not in the document]
Comment (nicholasperkins1992) [reply .env.example]: When we run it locally it requires the .env, the values.yaml will work when deploying with helm.

File: README.md [hunk 10]
```diff
@@ -129,21 +129,142 @@ Finally, run the replicator:
...
+```
+
+### Build Docker to Push to ACR
+
+Note: When you build the CDF Fabric Replicator from docker, the docker image is configured to use the config located in the Extraction Pipeline in CDF. You will need to configure the yaml file in the Extraction Pipeline in CDF.
+
+First, you should build your your docker image locally:
+
```
Comment (muralidharm) [inline README.md:169]: I usually have to run `az acr login` to auth to the container registry https://learn.microsoft.com/en-us/azure/container-registry/container-registry-authentication?tabs=azure-cli, might be worth adding as a step

Comment (nicholasperkins1992) [reply README.md:169]: Good idea, to keep this pr focused i will do it in a separate PR

File: cdf-fabric-replicator-chart/values.yaml [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+podAnnotations: {}
+
+podSecurityContext:
+  runAsUser: 1000
+  runAsGroup: 3000
+  fsGroup: 2000
+
+securityContext:
+  runAsNonRoot: true
+  readOnlyRootFilesystem: true
```
Comment (muralidharm) [inline cdf-fabric-replicator-chart/values.yaml:25]: Getting file permission errors as a result of adding these securityContext settings

Comment (nicholasperkins1992) [reply cdf-fabric-replicator-chart/values.yaml:25]: I tested this, the issue occurs with the dockerfile and the entrypoint.sh, we already have a fix but it is in the slim docker image. I am going to remove the security context in this pr and add it to the pr with the slim docker image

Comment (nicholasperkins1992) [reply cdf-fabric-replicator-chart/values.yaml]: Made a new PR to introduce this change https://github.com/cognitedata/cdf-fabric-replicator/pull/92
